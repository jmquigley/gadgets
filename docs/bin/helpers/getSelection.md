<a name="getSelectionShim"></a>

## getSelectionShim()
DOM Traversal is not implemented in JSDOM
The best we can do is shim the functions

**Kind**: global function  
