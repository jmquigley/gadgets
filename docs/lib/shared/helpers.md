## Functions

<dl>
<dt><a href="#debug">debug(context, args)</a></dt>
<dd><p>A wrapper for the debug function.  It uses the package.json to enable or
disable debugging information in the library.  It also turns off
coverage checks for this function.</p>
</dd>
<dt><a href="#globalize">globalize(name, pkg, replace)</a> ⇒ <code>object</code></dt>
<dd><p>Takes a variable name and an object and places that name and associated
object into the global node space.  If the object is alread in the
global space, then that reference is used and returned.  The replace
flag is used to override the current global reference.</p>
</dd>
</dl>

<a name="debug"></a>

## debug(context, args)
A wrapper for the debug function.  It uses the package.json to enable or
disable debugging information in the library.  It also turns off
coverage checks for this function.

**Kind**: global function  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>string</code> | the name of the module where this debug request is made. |
| args | <code>Array.&lt;object&gt;</code> | the list of arguments passed to the the debug function. |

<a name="globalize"></a>

## globalize(name, pkg, replace) ⇒ <code>object</code>
Takes a variable name and an object and places that name and associated
object into the global node space.  If the object is alread in the
global space, then that reference is used and returned.  The replace
flag is used to override the current global reference.

**Kind**: global function  
**Returns**: <code>object</code> - the global instance reference  

| Param | Type | Description |
| --- | --- | --- |
| name | <code>string</code> | the name of the variable to store globally |
| pkg | <code>object</code> | the object associated with the variable name |
| replace | <code>boolean</code> | if true, then it will always overwrite the current global (if it exists) |

