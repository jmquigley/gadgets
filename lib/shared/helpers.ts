"use strict";

const debug = require("debug")("helpers");

/**
 * Takes a variable name and an object and places that name and associated
 * object into the global node space.  If the object is already in the
 * global space, then that reference is used and returned.  The replace
 * flag is used to override the current global reference.
 * @param name {string} the name of the variable to store globally
 * @param pkg {object} the object associated with the variable name
 * @param replace {boolean} if true, then it will always overwrite the
 * current global (if it exists)
 * @return {object} the global instance reference
 */
export function globalize(
	name: string,
	pkgname: any,
	replace: boolean = false
) {
	let ref: any = pkgname;
	if (!(global as any)[name] || replace) {
		(window as any)[name] = (global as any)[name] = pkgname;
		debug("Adding global variable %o -> %O", name, (global as any)[name]);
	} else {
		ref = (global as any)[name];
	}

	return ref;
}
