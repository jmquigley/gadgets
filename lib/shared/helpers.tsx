'use strict';

const debugEntry = require('debug');
const pkg = require('../../package.json');

/**
 * A wrapper for the debug function.  It uses the package.json to enable or
 * disable debugging information in the library.  It also turns off
 * coverage checks for this function.
 * @param context {string} the name of the module where this debug request
 * is made.
 * @param args {object[]} the list of arguments passed to the the debug
 * function.
 */
export function debug(context: string, ...args: any[]) {
	/* istanbul ignore if  */
	if (pkg.debug) {
		const debugFn = debugEntry(`Gadgets -> ${context}`);
		debugFn(...args);
	}
}

/**
 * Takes a variable name and an object and places that name and associated
 * object into the global node space.  If the object is alread in the
 * global space, then that reference is used and returned.  The replace
 * flag is used to override the current global reference.
 * @param name {string} the name of the variable to store globally
 * @param pkg {object} the object associated with the variable name
 * @param replace {boolean} if true, then it will always overwrite the
 * current global (if it exists)
 * @return {object} the global instance reference
 */
export function globalize(name: string, pkgname: any, replace: boolean = false) {
	let ref: any = pkgname;
	if (!(global as any)[name] || replace) {
		(window as any)[name] = (global as any)[name] = pkgname;
	} else {
		ref = (global as any)[name];
	}

	return ref;
}
