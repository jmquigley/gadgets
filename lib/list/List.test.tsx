'use strict';

import {cleanup, header, log, mockupEnv} from '../../test/helpers';
mockupEnv();

import test from 'ava';
import {mount} from 'enzyme';
import * as path from 'path';
import * as React from 'react';
import {List} from './index';

test.before(t => {
	header(path.basename(__filename), t);
});

test.after.always.cb(t => {
	cleanup(path.basename(__filename), t);
});

test('Test the creation of a List control container', t => {
	const ctl = mount(
		<List>
			<li>some list item</li>
		</List>
	);

	t.truthy(ctl);
	log.debug(ctl.html(), __filename);

	t.regex(ctl.prop('id'), /[0-9a-zA-Z]{32}/);
	t.true(ctl.prop('enabled'));
	t.true(ctl.prop('visible'));

	t.is(ctl.find('.ui').length, 1);
	t.is(ctl.find('.ui-list').length, 1);
	t.is(ctl.find('.list').length, 1);
});

test('Test disabling of a List control', t => {
	const ctl = mount(
		<List enabled={false}>
			<li>some list item</li>
		</List>
	);

	t.truthy(ctl);
	log.debug(ctl.html(), __filename);

	t.regex(ctl.prop('id'), /[0-9a-zA-Z]{32}/);
	t.false(ctl.prop('enabled'));
	t.true(ctl.prop('visible'));

	t.is(ctl.find('.listDisabled').length, 1);
	t.is(ctl.find('.ui').length, 1);
	t.is(ctl.find('.ui-list').length, 1);
	t.is(ctl.find('.list').length, 1);
});

test('Test making List control invisible', t => {
	const ctl = mount(
		<List visible={false}>
			<li>some list item</li>
		</List>
	);

	t.truthy(ctl);
	log.debug(ctl.html(), __filename);

	t.regex(ctl.prop('id'), /[0-9a-zA-Z]{32}/);
	t.true(ctl.prop('enabled'));
	t.false(ctl.prop('visible'));

	t.is(ctl.find('.listInvisible').length, 1);
	t.is(ctl.find('.ui').length, 1);
	t.is(ctl.find('.ui-list').length, 1);
	t.is(ctl.find('.list').length, 1);
});
