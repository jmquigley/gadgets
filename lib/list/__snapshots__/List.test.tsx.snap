// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test disabling of a List control 1`] = `
<Wrapper
  alternating={false}
  backgroundColor={null}
  border={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  controlled={true}
  direction="right"
  disabled={true}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  id={null}
  left={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noborder={false}
  noedit={false}
  nohover={false}
  noselect={false}
  notheme={false}
  notooltip={false}
  obj="List"
  onError={[Function]}
  onSelect={[Function]}
  padding={null}
  reset={false}
  ripple={false}
  selected={false}
  sizing="normal"
  style={Object {}}
  testing={true}
  theme={null}
  tooltip={null}
  top={null}
  visible={true}
  width={null}
  xcss={null}
>
  <List__ListView
    alternating={false}
    className="ui-list nohover"
    style={Object {}}
  >
    <li
      href={
        Object {
          "selectHandler": [Function],
          "sizing": "normal",
        }
      }
      id="0"
      selected={false}
      sizing="normal"
    >
      some list item
    </li>
  </List__ListView>
</Wrapper>
`;

exports[`Test making List control invisible 1`] = `
<Wrapper
  alternating={false}
  backgroundColor={null}
  border={null}
  borderColor={null}
  bottom={null}
  className={null}
  color={null}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  id={null}
  left={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noborder={false}
  noedit={false}
  nohover={false}
  noselect={false}
  notheme={false}
  notooltip={false}
  obj="List"
  onError={[Function]}
  onSelect={[Function]}
  padding={null}
  reset={false}
  ripple={false}
  selected={false}
  sizing="normal"
  style={Object {}}
  testing={true}
  theme={null}
  tooltip={null}
  top={null}
  visible={false}
  width={null}
  xcss={null}
>
  <List__ListView
    alternating={false}
    className="ui-list"
    style={Object {}}
  >
    <li
      href={
        Object {
          "selectHandler": [Function],
          "sizing": "normal",
        }
      }
      id="0"
      selected={false}
      sizing="normal"
    >
      some list item
    </li>
  </List__ListView>
</Wrapper>
`;

exports[`Test retrieval of List props object 1`] = `
Object {
  "alternating": false,
  "backgroundColor": null,
  "border": null,
  "borderColor": null,
  "bottom": null,
  "children": null,
  "className": null,
  "color": null,
  "controlled": true,
  "direction": "right",
  "disabled": false,
  "err": null,
  "errorMessage": null,
  "focus": false,
  "height": null,
  "id": null,
  "left": null,
  "location": "none",
  "maxHeight": null,
  "maxWidth": null,
  "minHeight": null,
  "minWidth": null,
  "noborder": false,
  "noedit": false,
  "nohover": false,
  "noselect": false,
  "notheme": false,
  "notooltip": false,
  "obj": "List",
  "onSelect": [Function],
  "padding": null,
  "ripple": false,
  "selected": false,
  "sizing": "normal",
  "style": Object {},
  "testing": true,
  "theme": null,
  "tooltip": null,
  "top": null,
  "visible": true,
  "width": null,
  "xcss": null,
}
`;

exports[`Test retrieval of List props object 2`] = `
Object {
  "children": null,
  "keys": null,
  "selectedItem": null,
  "sizing": "normal",
  "style": Object {},
}
`;

exports[`Test the creation of a List control container 1`] = `
<Wrapper
  alternating={true}
  backgroundColor={null}
  border={null}
  borderColor={null}
  bottom={null}
  className="test-class"
  color={null}
  controlled={true}
  direction="right"
  disabled={false}
  err={null}
  errorMessage={null}
  focus={false}
  height={null}
  id={null}
  left={null}
  location="none"
  maxHeight={null}
  maxWidth={null}
  minHeight={null}
  minWidth={null}
  noborder={false}
  noedit={false}
  nohover={false}
  noselect={false}
  notheme={false}
  notooltip={false}
  obj="List"
  onError={[Function]}
  onSelect={[Function]}
  padding={null}
  reset={false}
  ripple={false}
  selected={false}
  sizing="normal"
  style={Object {}}
  testing={true}
  theme={null}
  tooltip={null}
  top={null}
  visible={true}
  width={null}
  xcss={null}
>
  <List__ListView
    alternating={true}
    className="ui-list test-class"
    style={Object {}}
  >
    <li
      href={
        Object {
          "selectHandler": [Function],
          "sizing": "normal",
        }
      }
      id="0"
      selected={false}
      sizing="normal"
    >
      some list item
    </li>
  </List__ListView>
</Wrapper>
`;
